odl/tomo/backends/astra_setup.py

```diff
@@ -0,0 +1,607 @@
+    # this was possible only for 2D geometries. Thus, standard ASTRA
+    # projection geometries do not have to be rescaled any more by the (
+    # isotropic) voxel size.
+    if isinstance(geometry, Parallel2dGeometry):
+        # TODO: change to parallel_vec when available
+        det_width = geometry.det_grid.stride[0]
+        det_count = geometry.detector.npixels
+        # convention in 'astra_conebeam_2d_geom_to_vec' differs from ASTRA's
+        angles = geometry.motion_grid.coord_vectors[0] - np.pi / 2
+        proj_geom = astra.create_proj_geom(
+            'parallel', det_width, det_count, angles)
+
+    elif (isinstance(geometry, DivergentBeamGeometry) and
+          isinstance(geometry.detector, FlatDetector) and
+          geometry.ndim == 2):
+        det_count = geometry.detector.npixels
+        vec = astra_conebeam_2d_geom_to_vec(geometry)
+        proj_geom = astra.create_proj_geom(
+            'fanflat_vec', det_count, vec)
+
+    elif (isinstance(geometry, ParallelGeometry) and
+          isinstance(geometry.detector, FlatDetector) and
+          geometry.ndim == 3):
+        det_row_count = geometry.det_grid.shape[1]
+        det_col_count = geometry.det_grid.shape[0]
+        vec = astra_parallel_3d_geom_to_vec(geometry)
+        proj_geom = astra.create_proj_geom(
+            'parallel3d_vec', det_row_count, det_col_count, vec)
+
+    elif (isinstance(geometry, DivergentBeamGeometry) and
+          isinstance(geometry.detector, FlatDetector) and
+          geometry.ndim == 3):
+        det_row_count = geometry.det_grid.shape[1]
+        det_col_count = geometry.det_grid.shape[0]
+        vec = astra_conebeam_3d_geom_to_vec(geometry)
+        proj_geom = astra.create_proj_geom(
+            'cone_vec', det_row_count, det_col_count, vec)
+    else:
+        raise NotImplementedError('unknown ASTRA geometry type {}.'.format(
+            geometry))
+
+    if 'astra' not in geometry.implementation_cache:
+        # Save computed value for later
+        geometry.implementation_cache['astra'] = proj_geom
+
+    return proj_geom
+
+
+def astra_data(astra_geom, datatype, data=None, ndim=2):
+    """Create an ASTRA data structure.
+
+    Parameters
+    ----------
+    astra_geom : `dict`
+        ASTRA geometry object for the data creator, must correspond to the
+        given data type
+    datatype : {'volume', 'projection'}
+        Type of the data container
+    data : `DiscreteLpVector`, optional
+        Data for the initialization of the data structure. If `None` creates
+        an ASTRA data object filled with zeros
+    ndim : {2, 3}, optional
+        Dimension of the data. If ``data`` is not `None`, this parameter
+        has no effect. Default: 2
+
+    Returns
+    -------
+    id : `int`
+        ASTRA internal ID for the new data structure
+    """
+    if data is not None:
+        if isinstance(data, DiscreteLpVector):
+            ndim = data.space.ndim
+        elif isinstance(data, np.ndarray):
```

Comment:
This option does not really make sense. I think we should only allow one datatype here (and I'd prefer DiscreteLpVector). This makes this function more simple. If we keep this, the docstring is wrong.